File: src/planner/lib/event-log.ts
// EventLog class: file I/O, heartbeat, serialization, and emit helpers.
// Extractors transform pi hook events into AuditEvent types.

import { promises as fs } from "node:fs";
import * as path from "node:path";
import type {
  AuditEvent,
  AuditEventPartial,
  HeartbeatEvent,
  PhaseStartEvent,
  StepTransitionEvent,
  PhaseEndEvent,
  Projection,
  ToolCallEvent,
  ToolResultEvent,
} from "./audit-events.js";
import { fold } from "./audit-fold.js";

// -- Pi event shapes (subset we consume) --

interface PiToolCallEvent {
  toolCallId: string;
  toolName: string;
  input: Record<string, unknown>;
}

interface PiToolResultEvent {
  toolCallId: string;
  toolName: string;
  input: Record<string, unknown>;
  content: Array<{ type: string; text?: string }>;
  isError: boolean;
}

// -- Constants --

const FILE_TOOLS = new Set(["read", "edit", "write"]);
const HEARTBEAT_MS = 10_000;

// Tools for which a bounded debug output preview is captured when debug mode
// is active. Intentionally narrow: only bash in this iteration.
const DEBUG_CAPTURE_TOOLS = new Set(["bash"]);

const DEBUG_CAPTURE_LIMIT = 4096;

// -- Helpers --

import { now } from "./time.js";

// -- Extractors --
// Transform pi's raw hook events into our audit event types.
// ts/seq are placeholders -- EventLog.append() overwrites them.

export function extractToolCall(piEvent: PiToolCallEvent): ToolCallEvent {
  return {
    kind: "tool_call",
    toolCallId: piEvent.toolCallId,
    tool: piEvent.toolName,
    input: piEvent.input,
    ts: now(),
    seq: 0,
  };
}

export function extractToolResult(
  piEvent: PiToolResultEvent,
  opts?: { debug?: boolean },
): ToolResultEvent {
  const { toolCallId, toolName, input, content, isError } = piEvent;

  const ev: ToolResultEvent = {
    kind: "tool_result",
    toolCallId,
    tool: toolName,
    error: isError,
    ts: now(),
    seq: 0,
  };

  // Capture output size for file and bash tools.
  if (FILE_TOOLS.has(toolName) && !isError) {
    const text = content.find((c) => c.type === "text")?.text ?? "";
    ev.lines = text.split("\n").length;
    ev.chars = text.length;
  } else if (toolName === "bash") {
    const text = content.find((c) => c.type === "text")?.text ?? "";
    ev.lines = text.split("\n").length;
    ev.chars = text.length;
  }

  // Preserve koan tool response text for projection use (completionSummary).
  if (toolName.startsWith("koan_")) {
    ev.koanResponse = content
      .filter((c) => c.type === "text" && c.text !== undefined)
      .map((c) => c.text as string);
  }

  // Debug mode: capture a bounded preview of tool output for designated tools.
  // Only populated when debug is active; never written in normal mode.
  // NOT folded into Projection — debug-only; never add to Projection.
  if (opts?.debug && DEBUG_CAPTURE_TOOLS.has(toolName) && !isError) {
    const text = content.find((c) => c.type === "text")?.text ?? "";
    ev.debugOutput =
      text.slice(0, DEBUG_CAPTURE_LIMIT) +
      (text.length > DEBUG_CAPTURE_LIMIT ? "\n\u2026[truncated]" : "");
  }

  void input; // suppress unused-variable warning (input is part of the public API shape)

  return ev;
}

// -- EventLog --

export class EventLog {
  private readonly eventsPath: string;
  private readonly statePath: string;
  private readonly stateTmpPath: string;
  private fd: fs.FileHandle | null = null;
  private seq = 0;
  private projection: Projection;
  private heartbeat: ReturnType<typeof setInterval> | null = null;
  // Serializes append() calls. Heartbeat timer and tool_result handler
  // both call append() concurrently -- without serialization, two
  // writeState() calls race on the shared tmp file (ENOENT on rename).
  private pending: Promise<void> = Promise.resolve();

  constructor(dir: string, role: string, phase: string, model: string | null = null) {
    this.eventsPath = path.join(dir, "events.jsonl");
    this.statePath = path.join(dir, "state.json");
    this.stateTmpPath = path.join(dir, "state.tmp.json");
    this.projection = {
      role,
      phase,
      model,
      status: "running",
      step: 0,
      totalSteps: 0,
      stepName: "",
      lastAction: null,
      currentToolCallId: null,
      updatedAt: now(),
      eventCount: 0,
      error: null,
      completionSummary: null,
      tokensSent: 0,
      tokensReceived: 0,
      lastToolResultAt: null,
    };
  }

  async open(): Promise<void> {
    this.fd = await fs.open(this.eventsPath, "a");
    await this.writeState();
    // Heartbeat keeps updatedAt fresh even during long-running steps.
    // unref() so the timer doesn't prevent process exit — pi's print mode
    // relies on natural event loop drain (no process.exit()) and never
    // emits session_shutdown, so EventLog.close() may not be called.
    this.heartbeat = setInterval(() => {
      void this.append({ kind: "heartbeat" } as Omit<HeartbeatEvent, "ts" | "seq">);
    }, HEARTBEAT_MS);
    this.heartbeat.unref();
  }

  // Assigns ts + seq, appends JSON line, folds, writes state atomically.
  // Serialized: concurrent callers queue behind the in-flight write.
  async append(partial: AuditEventPartial): Promise<void> {
    const task = () => this.doAppend(partial);
    this.pending = this.pending.then(task, task);
    return this.pending;
  }

  private async doAppend(partial: AuditEventPartial): Promise<void> {
    if (!this.fd) {
      throw new Error("EventLog.append called before open()");
    }

    const e = { ...partial, ts: now(), seq: this.seq++ } as AuditEvent;
    await this.fd.write(JSON.stringify(e) + "\n");
    this.projection = fold(this.projection, e);
    await this.writeState();
  }

  async emitPhaseStart(totalSteps: number): Promise<void> {
    await this.append({
      kind: "phase_start",
      phase: this.projection.phase,
      role: this.projection.role,
      model: this.projection.model,
      totalSteps,
    } as Omit<PhaseStartEvent, "ts" | "seq">);
  }

  async emitStepTransition(step: number, name: string, totalSteps: number): Promise<void> {
    await this.append({
      kind: "step_transition",
      step,
      name,
      totalSteps,
    } as Omit<StepTransitionEvent, "ts" | "seq">);
  }

  async emitPhaseEnd(outcome: "completed" | "failed", detail?: string): Promise<void> {
    await this.append({
      kind: "phase_end",
      outcome,
      detail,
    } as Omit<PhaseEndEvent, "ts" | "seq">);
  }



  async close(): Promise<void> {
    if (this.heartbeat) {
      clearInterval(this.heartbeat);
      this.heartbeat = null;
    }
    if (this.fd) {
      await this.fd.close();
      this.fd = null;
    }
  }

  get state(): Readonly<Projection> {
    return this.projection;
  }

  // Atomic write: tmp file then rename so readers never see partial JSON.
  private async writeState(): Promise<void> {
    const json = JSON.stringify(this.projection, null, 2) + "\n";
    await fs.writeFile(this.stateTmpPath, json);
    await fs.rename(this.stateTmpPath, this.statePath);
  }
}

// -- Exports --

// Reads state.json as a Projection; returns null if missing or malformed.
// Used by web server polling loop.
export async function readProjection(dir: string): Promise<Projection | null> {
  try {
    const raw = await fs.readFile(path.join(dir, "state.json"), "utf8");
    return JSON.parse(raw) as Projection;
  } catch {
    return null;
  }
}

File: koan/audit/event_log.py
# EventLog -- append-only audit trail with asyncio queue serialization.
# Python port of src/planner/lib/event-log.ts.
# Writes events.jsonl and state.json atomically to a subagent directory.

from __future__ import annotations

import asyncio
import json
import os
from dataclasses import asdict
from datetime import datetime, timezone
from pathlib import Path
from typing import TYPE_CHECKING, Any

import aiofiles

from .events import Projection, RunnerDiagnosticEvent
from .fold import fold

if TYPE_CHECKING:
    from ..runners.base import RunnerDiagnostic


# -- Helpers -------------------------------------------------------------------

def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


_SENTINEL = object()

HEARTBEAT_INTERVAL = 10.0


# -- EventLog ------------------------------------------------------------------

class EventLog:
    def __init__(self, subagent_dir: str, role: str, phase: str, model: str | None = None):
        self._dir = subagent_dir
        self._events_path = str(Path(subagent_dir) / "events.jsonl")
        self._state_path = str(Path(subagent_dir) / "state.json")
        self._state_tmp_path = str(Path(subagent_dir) / "state.tmp.json")
        self._seq = 0
        self._projection = Projection(
            role=role,
            phase=phase,
            model=model,
            status="running",
            updated_at=_now(),
        )
        self._queue: asyncio.Queue = asyncio.Queue()
        self._consumer_task: asyncio.Task | None = None
        self._fd: Any = None
        self._heartbeat_task: asyncio.Task | None = None

    async def open(self) -> None:
        Path(self._dir).mkdir(parents=True, exist_ok=True)
        self._fd = await aiofiles.open(self._events_path, "a")
        await self._write_state()
        self._consumer_task = asyncio.create_task(self._consume())
        self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())

    async def _heartbeat_loop(self) -> None:
        try:
            while True:
                await asyncio.sleep(HEARTBEAT_INTERVAL)
                await self.append({"kind": "heartbeat"})
        except asyncio.CancelledError:
            pass

    async def _consume(self) -> None:
        while True:
            item = await self._queue.get()
            if item is _SENTINEL:
                self._queue.task_done()
                break
            try:
                await self._do_append(item)
            finally:
                self._queue.task_done()

    async def _do_append(self, partial: dict) -> None:
        if self._fd is None:
            raise RuntimeError("EventLog.append called before open()")

        partial["ts"] = _now()
        partial["seq"] = self._seq
        self._seq += 1

        line = json.dumps(partial, default=str) + "\n"
        await self._fd.write(line)
        await self._fd.flush()

        # Reconstruct as a typed event for fold.
        # fold() dispatches on kind and reads attributes directly.
        from . import events as ev
        kind = partial.get("kind", "")
        event_cls = _KIND_MAP.get(kind)
        if event_cls is not None:
            event = event_cls(**{k: v for k, v in partial.items() if k in event_cls.__dataclass_fields__})
        else:
            return

        self._projection = fold(self._projection, event)
        await self._write_state()

    async def append(self, partial: dict) -> None:
        await self._queue.put(partial)

    async def emit_phase_start(self, total_steps: int) -> None:
        await self.append({
            "kind": "phase_start",
            "phase": self._projection.phase,
            "role": self._projection.role,
            "model": self._projection.model,
            "total_steps": total_steps,
        })

    async def emit_step_transition(self, step: int, name: str, total_steps: int) -> None:
        await self.append({
            "kind": "step_transition",
            "step": step,
            "name": name,
            "total_steps": total_steps,
        })

    async def emit_phase_end(self, outcome: str, detail: str | None = None) -> None:
        await self.append({
            "kind": "phase_end",
            "outcome": outcome,
            "detail": detail,
        })

    async def emit_runner_diagnostic(self, diag: RunnerDiagnostic) -> None:
        await self.append({
            "kind": "runner_diagnostic",
            "code": diag.code,
            "runner": diag.runner,
            "stage": diag.stage,
            "message": diag.message,
            "details": diag.details,
        })

    async def close(self) -> None:
        if self._heartbeat_task is not None:
            self._heartbeat_task.cancel()
            try:
                await self._heartbeat_task
            except asyncio.CancelledError:
                pass
            self._heartbeat_task = None

        if self._consumer_task is not None:
            await self._queue.put(_SENTINEL)
            await self._consumer_task
            self._consumer_task = None

        if self._fd is not None:
            await self._fd.close()
            self._fd = None

    @property
    def state(self) -> Projection:
        return self._projection

    async def _write_state(self) -> None:
        data = asdict(self._projection)
        content = json.dumps(data, indent=2) + "\n"
        async with aiofiles.open(self._state_tmp_path, "w") as f:
            await f.write(content)
        os.rename(self._state_tmp_path, self._state_path)


# -- Kind -> event class map ---------------------------------------------------

from .events import (
    HeartbeatEvent,
    PhaseEndEvent,
    PhaseStartEvent,
    RunnerDiagnosticEvent,
    StepTransitionEvent,
    ThinkingEvent,
    ToolCallEvent,
    ToolResultEvent,
    UsageEvent,
)

_KIND_MAP: dict[str, type] = {
    "phase_start": PhaseStartEvent,
    "step_transition": StepTransitionEvent,
    "phase_end": PhaseEndEvent,
    "heartbeat": HeartbeatEvent,
    "usage": UsageEvent,
    "thinking": ThinkingEvent,
    "tool_call": ToolCallEvent,
    "tool_result": ToolResultEvent,
    "runner_diagnostic": RunnerDiagnosticEvent,
}
